.github/workflows/rust.yml

```diff
@@ -5,21 +5,91 @@ on:
     branches: ["main"]
   pull_request:
     branches: ["main"]
+  workflow_dispatch:  # Allow manual triggering
 
 env:
   CARGO_TERM_COLOR: always
 
 jobs:
-  build:
+  lint:
     runs-on: ubuntu-latest
-
     steps:
       - uses: actions/checkout@v4
       - name: Run cargo-fmt
         run: cargo fmt --all -- --check
       - name: Run cargo-clippy
         run: cargo clippy -- -D warnings
-      - name: Build
-        run: cargo build --verbose
+      - name: Check
+        run: cargo check --verbose --workspace --all-targets
+
+  # Library tests (run in parallel with "tests" and dpcs)
+  test-lib:
+    runs-on: ubuntu-latest
+    needs: lint
+    steps:
+      - uses: actions/checkout@v4
+      - name: Run Tests
+        run: cargo test --lib --workspace
+
+  # "test" tests (aka tests in the tests directory)
+  test-test:
+    runs-on: ubuntu-latest
+    needs: lint
+    steps:
+      - uses: actions/checkout@v4
+      - name: Run Tests
+        run: cargo test --tests --workspace
+
+  # doc examples test
+  test-doc:
+    runs-on: ubuntu-latest
+    needs: lint
+    steps:
+      - uses: actions/checkout@v4
       - name: Run Tests
-        run: cargo test --verbose
+        run: cargo test --doc --workspace
+
+  # documentation build
+  docs:
+    runs-on: ubuntu-latest
+    needs: lint
+    steps:
+      - uses: actions/checkout@v4
+      - name : Build docs
+        env:
+          RUSTDOCFLAGS: "-D warnings"
+        run: cargo doc --no-deps --workspace
+
+
+  # Conformance tests (compare to existing  implementation)
+  conformance-test:
```

Comment:
Since this one takes a bit I gated it on the simpler unit tests passing so we don't waste builder time on the expensive one if the CI will fail on cheaper tests.